unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs

```diff
@@ -0,0 +1,311 @@
+            };
+            joint.slerpDrive = jd;
+        }
+
+    }
+
+    /// <summary>
+    /// Create BodyPart object and add it to dictionary.
+    /// </summary>
+    public void SetupBodyPart(Transform t)
+    {
+        BodyPart bp = new BodyPart
+        {
+            rb = t.GetComponent<Rigidbody>(),
+            joint = t.GetComponent<ConfigurableJoint>(),
+            startingPos = t.position,
+            startingRot = t.rotation
+        };
+		bp.rb.maxAngularVelocity = 100;
+        bodyParts.Add(t, bp);
+        bp.groundContact = t.GetComponent<CrawlerContact>();
+		bp.agent = this;
+    }
+    public override void InitializeAgent()
+    {
+		academy = FindObjectOfType<CrawlerAcademy>();
+        SetupBodyPart(body);
+        SetupBodyPart(leg0_upper);
+        SetupBodyPart(leg0_lower);
+        SetupBodyPart(leg1_upper);
+        SetupBodyPart(leg1_lower);
+        SetupBodyPart(leg2_upper);
+        SetupBodyPart(leg2_lower);
+        SetupBodyPart(leg3_upper);
+        SetupBodyPart(leg3_lower);
+		bodyParts[leg0_lower].rb.centerOfMass = footCenterOfMassShift;
+		bodyParts[leg1_lower].rb.centerOfMass = footCenterOfMassShift;
+		bodyParts[leg2_lower].rb.centerOfMass = footCenterOfMassShift;
+		bodyParts[leg3_lower].rb.centerOfMass = footCenterOfMassShift;
+    }
+
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void CollectObservationBodyPart(BodyPart bp)
+    {
+        var rb = bp.rb;
+        AddVectorObs(bp.groundContact.touchingGround ? 1 : 0); // Is this bp touching the ground
+
+        AddVectorObs(rb.velocity);
+        AddVectorObs(rb.angularVelocity);
+        Vector3 localPosRelToBody = body.InverseTransformPoint(rb.position);
+        AddVectorObs(localPosRelToBody);
+
+        if(bp.rb.transform != body)
+        {
+            AddVectorObs(Quaternion.FromToRotation(body.forward, bp.rb.transform.forward));
+        }
+    }
+
+    void Update()
+    {
+        //update pos to target every tick
+		dirToTarget = target.position - bodyParts[body].rb.position;
+
+        //if enabled the feet will light up green when the foot is grounded.
+        //this is just a visualization and isn't necessary for function
+        if(useFootGroundedVisualization)
+        {
+            foot0.material = bodyParts[leg0_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot1.material = bodyParts[leg1_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot2.material = bodyParts[leg2_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot3.material = bodyParts[leg3_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+        }
+    }
+
+	/// <summary>
+    /// Adds the raycast hit dist and relative pos to observations
+    /// </summary>
+    void RaycastObservation(Vector3 pos, Vector3 dir, float maxDist)
+    {
```

Comment:
Is it possible to only use the Raycast component introduced in 0.3 ? Why do you need a new Raycast method ?

Reply:
this one is much less complex and also gets the hit position relative to the body